xilinx/xdc.cc

```diff
@@ -115,69 +121,73 @@ void Arch::parseXdc(std::istream &in)
+                    log_error("expected four arguments to 'set_property' (on line %d)\n", lineno);
+                else if (arguments.at(1) == "-dict") {
+                    std::vector<std::string> dict_args = split_to_args(strip_quotes(arguments.at(2)), false);
+                    if ((dict_args.size() % 2) != 0)
+                        log_error("expected an even number of argument for dictionary (on line %d)\n", lineno);
+                    arg_pairs.reserve(dict_args.size() / 2);
+                    for (int cursor = 0; cursor + 1 < int(dict_args.size()); cursor += 2) {
+                        arg_pairs.emplace_back(std::move(dict_args.at(cursor)), std::move(dict_args.at(cursor + 1)));
+                    }
                 } else
-                    break;
-            }
-            if (!got_period)
-                log_error("found create_clock without period (on line %d)", lineno);
-            std::vector<NetInfo *> dest = get_nets(arguments.at(cursor));
-            for (auto n : dest) {
-                n->clkconstr = std::unique_ptr<ClockConstraint>(new ClockConstraint);
-                n->clkconstr->period = DelayPair(getDelayFromNS(period));
-                n->clkconstr->high = DelayPair(getDelayFromNS(period / 2));
-                n->clkconstr->low = DelayPair(getDelayFromNS(period / 2));
+                    arg_pairs.emplace_back(std::move(arguments.at(1)), std::move(arguments.at(2)));
+                if (arguments.at(1) == "INTERNAL_VREF")
+                    continue;
+                if (arguments.at(3).size() > 2 && arguments.at(3) == "[current_design]") {
+                    log_warning("[current_design] isn't supported, ignoring (on line %d)\n", lineno);
+                    continue;
+                }
+                std::vector<CellInfo *> dest = get_cells(arguments.at(3));
+                for (auto c : dest)
+                    for (const auto &pair : arg_pairs)
+                        c->attrs[id(pair.first)] = std::string(pair.second);
+            } else if (cmd == "create_clock") {
+                double period = 0;
+                bool got_period = false;
+                int cursor = 1;
+                for (cursor = 1; cursor < int(arguments.size()); cursor++) {
+                    std::string opt = arguments.at(cursor);
+                    if (opt == "-add")
+                        ;
+                    else if (opt == "-name" || opt == "-waveform")
+                        cursor++;
+                    else if (opt == "-period") {
+                        cursor++;
+                        period = std::stod(arguments.at(cursor));
+                        got_period = true;
+                    } else
+                        break;
+                }
+                if (!got_period)
+                    log_error("found create_clock without period (on line %d)", lineno);
+                std::vector<NetInfo *> dest = get_nets(arguments.at(cursor));
```

Comment:
Can you put in the logic for cleaning out spaces and quotes here? This should be legal:
```
create_clock -period 10.0 -name sys_clk_pin -waveform {0.000 5.000} -add [get_ports { sys_clk_i } ]
```
In other words, fix `get_nets()` similar to how you fixed `get_cells()`.

Reply:
@gh-sh4 I think if we have addressed this issue, we can merge it. What do you think?

Reply:
@hansfbaier Feel free to merge this. I agree, it would be great if the case @engstad mentioned would be handled, but I think in reality, this parser should be slightly re-written to better support these things, which could be another (probably low-priority) issue. We shouldn't need to duplicate this logic in multiple places. Currently to address engstad's issue, a user must avoid create_clock with `{}` style get_ports.

Reply:
OK will merge. Thanks!